### ISSUE TITLE
Project status?

### ISSUE THREAD

**Issue by tamasgal** [recipy/recipy#219]
I am thinking about using this project but I found that it only works with an older tinydb (and serializations).

What is the actual project status? Are PR still welcome (and accepted) or is there another successor of recipy which I have not found yet? :wink:

**Comment by robintw** [comment]
Hi @tamasgal - unfortunately work has stalled on recipy for quite a while. I was hoping to get funding to continue work on it (I'm now a freelance developer/data scientist), but haven't been successful. I think most of it still works, but you're right that it will depend on older versions of libraries.

Unfortunately I haven't found any other projects doing quite what recipy does/did.

**Comment by jvdzwaan** [comment]
It is really too bad the project isn't active anymore. My (spare) time is also limited...

By coincidence I recently came across https://mlflow.org which allows tracking things in Python code. There is no automatic tracking of files, though.

**Comment by tamasgal** [comment]
@robintw I might help out and update the dependencies including the modifications needed to make it work.

@jvdzwaan thanks for the link, that's interesting

**Comment by tamasgal** [comment]
@robintw I'd like to continue working on this project starting from the second half of August. Would you mind adding me as a project member?

**Comment by robintw** [comment]
Sure, I've added you as a collaborator. It's great that you want to get involved!

I may also have a bit of time in August, so it'd be good to discuss what specific work you'd like to do, and how that links in with our original plans for recipy and the sort of things I might have chance to work on.

Do you want to start creating issues for the things you want to work on, or prioritise the existing issues in the [Project Board](https://github.com/recipy/recipy/projects/1)?

**Comment by tamasgal** [comment]
Thanks!

Yes, so mainly I'd like to update the dependencies, get the test suite running again and clean up some things first, also to understand the core concepts.

I'd like to integrate this tool into our processing pipelines to do automatic provenance in scientific analysis and productions. I like the overall concept and I think there is a large target audience.

Yes I'll create issues and PRs for everything I come up with.

**Comment by pravsripad** [comment]
Hello, first I'd like to thank the developers for this project. I find it amazing because it is so easy to use - one import and bam! provenance tracking ! 

However I'd like to ask the first question by @tamasgal once again? Is the project being maintained / are PRs welcome? I see some commits / PRs on and off and thought I'd ask.

**Comment by pravsripad** [comment]
Have the developers considered/tried seeking funds from sources that support scientifically relevant open source projects? Heck, maybe even crowd funding :)

**Comment by robintw** [comment]
Hi @pravsripad. Thanks for your interest in recipy. I was one of the original authors, and agree that there is a lot of potential here. However, all the original authors have other day jobs, and we have never managed to find any funding to continue work on it. Most of the funding sources for scientific open source development require proof of a community already in place around the software, and demand from the scientific community - and we've never managed to get that.

I now work as a freelance software engineer and data scientist (see www.rtwilson.com), and have quite a lot of paid work to do at the moment. Having said that, I'd be happy to do more work on this in the future if I could get paid for it - but I don't know how we can get enough funding for that.

If you're aware of any available funding or want to get involved in getting funding to work on this then I'd be happy to discuss this with you.

So, in summary: I agree that this project is great (well, I would, wouldn't I!), but we'll need some proper funding to progress further with it.

**Comment by tamasgal** [comment]
I am really sorry, things went a bit differently than expected, unfortunately. Our data processing pipelines have a huge heterogenous stack of software (different languages and frameworks) and the steering of those is also not exclusively Python, so that we were kind of forced to rethink the provenance handling and now have a mix of external/internal provenance tracking.

`recipy` is definitely a very unique and outstanding solution for Python-based workflows, so I'd love to see this project rising again but my time is now fully dedicated to other projects, so I cannot help out.

**Comment by pravsripad** [comment]
Thank you both for your quick reply. I completely understand that it gets difficult as life goes in different ways. It definitely is a pity to let this nice project go quiet. 

@robintw I'll keep a watch out and let you know if I come across any sources for funding this and spread the word with my colleagues. You do accept PRs still though right? Is there a paper / conference report that can be cited along with this package? 

@tamasgal if you found any similar tools for provenance tracking, I'm all ears :)

**Comment by pravsripad** [comment]
@robintw nice blog !

**Comment by tamasgal** [comment]
@pravsripad we use [Nextflow](https://www.nextflow.io) and [SnakeMake](https://snakemake.readthedocs.io) for complex production pipelines in our scientific workflows and those tools generate provenance information for input, output files and the activities. We also have many internal provenance formats which are generated by various frameworks separately, e.g. by [thepipe](https://thepipe.readthedocs.io/en/latest/) which has integrated provenance tracking, very similar to `recipy` but not fully automated. `thepipe` is then used in Python-based scripts to create modular pipelines and its provenance tracking is utilised in many of our Python-based frameworks/libraries to mark activities, input and output files.

It's a fairly complex topic. If everything would be Python based, `recipy` would cover most of the cases, but the reality is very different, unfortunately. We also start introducing Julia (IMHO the future language for scientific computing) and that would need yet another internal provenance facility, which I am also working on currently.

So yes, as said, it's a very complex topic and ther is no general solution. It all depends on your needs and requirements which should be defined beforehand as clearly as possible. This was also one of our problems: requirements were not communicated clearly.

**Comment by pravsripad** [comment]
@tamasgal Yes, it can get quite complicated indeed. I am rooting for `recipy` because we use primarily python, and it does not require a major change in the code / workflow for all the researchers using it. Some may not be very well versed with coding also. Further all of the data are in json essentially making it future proof, something that is important for reproducibility. 

Do you know if there exist any resources that cover all such tracking methods? Maybe a review paper?

**Comment by tycho-kirchner** [comment]
> Do you know if there exist any resources that cover all such tracking methods? Maybe a review paper?

Please excuse the self promotion, but you may be interested in our tool [shournal](https://github.com/tycho-kirchner/shournal), a file-tracker for the shell. It tracks files by tracing the operations way down in the Linux-kernel - thus, it works irrespective of the programming language. There's also a preprint [Bashing irreproducibility with shournal](https://doi.org/10.1101/2020.08.03.232843) on bioRxiv. Thanks, Tycho

**Comment by pravsripad** [comment]
> > Do you know if there exist any resources that cover all such tracking methods? Maybe a review paper?
> 
> Please excuse the self promotion, but you may be interested in our tool [shournal](https://github.com/tycho-kirchner/shournal), a file-tracker for the shell. It tracks files by tracing the operations way down in the Linux-kernel - thus, it works irrespective of the programming language. There's also a preprint [Bashing irreproducibility with shournal](https://doi.org/10.1101/2020.08.03.232843) on bioRxiv. Thanks, Tycho

Thank you, I will have a look.
